libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetector.java

```diff
@@ -226,24 +224,28 @@ private static boolean movingAwayFromManeuver(RouteProgress routeProgress,
     LineString remainingStepLineString = TurfMisc.lineSlice(userPointOnStep, maneuverPoint, stepLineString);
     double userDistanceToManeuver = TurfMeasurement.length(remainingStepLineString, TurfConstants.UNIT_METERS);
 
-    boolean hasDistances = !distancesAwayFromManeuver.isEmpty();
-    boolean validOffRouteDistanceTraveled = hasDistances && distancesAwayFromManeuver.peekLast()
-      - distancesAwayFromManeuver.peekFirst() < MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE;
-    boolean exceedsManeuverDistancesThreshold = validOffRouteDistanceTraveled
-      && distancesAwayFromManeuver.size() >= 3;
-
-    if (exceedsManeuverDistancesThreshold) {
-      // User's moving away from maneuver position, thus offRoute.
-      return true;
-    }
     if (distancesAwayFromManeuver.isEmpty()) {
-      distancesAwayFromManeuver.push((int) userDistanceToManeuver);
-    } else if (userDistanceToManeuver > distancesAwayFromManeuver.peek()) {
-      distancesAwayFromManeuver.push((int) userDistanceToManeuver);
-    } else {
-      // If we get a descending distance, reset the counter
+      // No move-away positions before, add the current one to history stack
+      distancesAwayFromManeuver.add((int) userDistanceToManeuver);
+    } else if ((int) userDistanceToManeuver > distancesAwayFromManeuver.getLast()) {
+      // If distance to maneuver increased (wrong way), add new position to history stack
+
+      if (distancesAwayFromManeuver.size() >= 3) {
+        // We replace the first one, in order to keep the history with the last one
+        distancesAwayFromManeuver.removeLast();
+      }
+      distancesAwayFromManeuver.add((int) userDistanceToManeuver);
+    } else if ((int) userDistanceToManeuver < distancesAwayFromManeuver.getLast()) {
+      // If distance to maneuver decreased (right way) clean history
```

Comment:
How about we add a threshold of a few meters here? For example

`userDistanceToManeuver < distancesAwayFromManeuver.getLast() - X`

X could be 5m for example, just to avoid resetting too easily due to inaccuracy in the GPS signal?

Reply:
Yes, a threshold for this sounds good! 

Do we need to add this to the navigation options as well? I think it's great to be able to configure everything, but this seems like too much. What do you think?

Reply:
May be we can use here the same value like `MINIMUM_BACKUP_DISTANCE_FOR_OFF_ROUTE` or even a dynamic option.

Reply:
I think adding this to the nav options would be great, I didn't want to ask too much of you, but yes, we should allow to configure as much as possible for these values, as every app and use case need their own specific values. I am open on how we do this, probably adding a new option would be best to avoid unexpected breaking changes?